spring-pulsar/src/main/java/org/springframework/pulsar/listener/PulsarMessageRecoverer.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.pulsar.listener;
+
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+import org.apache.pulsar.client.api.Consumer;
+import org.apache.pulsar.client.api.Message;
+
+/**
+ * Allows recovering a failed Pulsar message.
+ *
+ * Implementations can choose how the message needs to be recovered by providing a
+ * {@link java.util.function.Function} implementation that takes a {@link Consumer} and
+ * then provide a {@link BiConsumer} which takes {@link Message} and the thrown
+ * {@link Exception}.
+ *
+ * @param <T> payload type of Pulsar message.
+ * @author Soby Chacko
+ */
+@FunctionalInterface
+public interface PulsarMessageRecoverer<T> extends Function<Consumer<T>, BiConsumer<Message<T>, Exception>> {
```

Comment:
I know this was inspired by the reocoverer in `spring-kafka` of which I used quite a bit at my previous place. I found this shortcut of using functions/bi-consumer to loosely type a factory->impl contract less than ideal. At 1st glance it is really hard for users to resonate on what they are supposed to do w/ a Function->BiConsumer. Nothing really says that the BiConsumer is a message recoverer. Its not really, its a biconsumer. 

I would instead do something more traditional such as this: 

```java
@FunctionalInterface
public interface PulsarMessageRecovererFactory<T> {
	PulsarMessageRecoverer<T> recovererForConsumer(Consumer<T> consumer);
}

@FunctionalInterface
public interface PulsarMessageRecoverer<T> {
	void recoverMessage(Message<T> message, Exception exception);
}

public class PulsarDeadLetterPublishingRecoverer<T> implements PulsarMessageRecovererFactory<T> {
	@Override
	public PulsarMessageRecoverer<T> recoverForConsumer(Consumer<T> consumer) {
		return (message, exception) -> {
			try {
				pulsarTemplate.newMessage(message.getValue())
						.withTopic(destinationResolver.apply(consumer, message))
						.withMessageCustomizer(messageBuilder -> messageBuilder.property(EXCEPTION_THROWN_CAUSE,
								exception.getCause().getMessage()))
						.sendAsync();
			}
			catch (PulsarClientException e) {
				logger.error(e, "DLT publishing failed.");
			}
		};
	}
}
```

This makes sense to me and is easy for me to understand what I need to do/impl to create my own. 

Wdyt?

Reply:
Excellent!